### PR TITLE
fcntl: set FD_CLOEXEC on file descriptors

### PR DESCRIPTION
**Pull request by rittneje**
Set `FD_CLOEXEC` on file descriptors. This prevents child processes from inheriting them, which is almost always the intended behavior.

Partial fix for #3671. For simplicity it only fixes usages of direct file descriptors. It does not fix usages of `popen` or `fopen`.

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [N/A] Example configuration file for the change
- [N/A] Debug log output from testing the change
- [N/A] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**
- [N/A] Documentation required for this feature

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Comment by rittneje** [comment]
cc @koleini

**Comment by rittneje** [comment]
cc @nokute78

**Comment by rittneje** [comment]
ping @edsiper @koleini @nokute78

**Comment by rittneje** [comment]
@edsiper @koleini @nokute78 Please take a look.

**Comment by rittneje** [comment]
@edsiper @koleini @nokute78 Please take a look.

**Comment by rittneje** [comment]
@edsiper @koleini @nokute78 Please review. This PR has been open for almost two months.

**Comment by edsiper** [comment]
why do we need it ?, I would like to understand the use case: we are not using execve(2)

**Comment by rittneje** [comment]
@edsiper `execvp` is used here. https://github.com/fluent/fluent-bit/blob/38ad394fb1950deb9e383060223e3f853aa7c1eb/src/aws/flb_aws_credentials_process.c#L503

And in any case using `O_CLOEXEC` all the time in the first place is a best practice. Otherwise you have to go back and add it everywhere once you do need to use any `exec` functions.

**Comment by rittneje** [comment]
Not stale. @edsiper @koleini @nokute78 Please review. This PR has been open for three months.

**Comment by rittneje** [comment]
Not stale.

**Comment by rittneje** [comment]
@edsiper @koleini @nokute78 Please review. This PR has been open for almost 7 months.

**Comment by PettitWesley** [comment]
@rittneje Please forgive me because I don't know things, are there any side effects that can occur with setting this flag? I am wondering if we need to go through and test all of this code which could take a while, or if its safe to accept unless there are specific concerns?

(Also this one will definitely need approval from one of the other maintainers, I can't approve/merge this myself).

**Comment by rittneje** [comment]
@PettitWesley This flag (`O_CLOEXEC` and friends) instructs the kernel to automatically close the file descriptor when you call `exec` and friends. It will not have any effect in any other context.

One place that `exec` is used is in the AWS credential process logic I added last year. In this case only stdin/stdout/stderr matter, and the flag is not being applied to these fds. Plus the tests are passing, so it should be fine.
https://github.com/fluent/fluent-bit/blob/9bb0c1b2980bd5d63d98a628189444f42d7047f5/src/aws/flb_aws_credentials_process.c#L503

The only other occurrence I see is in the monkey library. I don't really understand what this code is doing or where it is used, but from what I can tell, it is similar to my credential process logic in that only stdin/stdout/stderr matter.
https://github.com/fluent/fluent-bit/blob/9bb0c1b2980bd5d63d98a628189444f42d7047f5/lib/monkey/plugins/cgi/cgi.c#L281-L288

**Comment by PettitWesley** [comment]
@rittneje Thanks. I do not see a concern with this patch, however, some other higher ranking maintainer like @edsiper or @nokute78 will have to approve as well.

**Review by PettitWesley — DISMISSED** [review]
Seems like the right thing to me, but requires approval from true experts.

**Comment by rittneje** [comment]
@edsiper @niedbalski @koleini @fujimotos @nokute78

**Comment by rittneje** [comment]
@edsiper @niedbalski @koleini @fujimotos @nokute78

**Comment by rittneje** [comment]
@edsiper @niedbalski @koleini @fujimotos @nokute78

**Comment by PettitWesley** [comment]
> Set FD_CLOEXEC on file descriptors. This prevents child processes from inheriting them, which is almost always the intended behavior.

Wait- Fluent Bit doesn't spin up any child processes though right? In practice, how would this happen?

**Comment by rittneje** [comment]
@PettitWesley Yes it does.
https://github.com/fluent/fluent-bit/blob/38ad394fb1950deb9e383060223e3f853aa7c1eb/src/aws/flb_aws_credentials_process.c#L503

And even if it didn't, it is always a best practice use use `O_CLOEXEC` etc. from the start, or else it will become an issue when you do start exec'ing things.

**Comment by PettitWesley** [comment]
@rittneje I see, I think probably this input does as well: https://docs.fluentbit.io/manual/pipeline/inputs/exec

And also the new `Kube_Token_Command` option in the k8s filter does too IIRC...

**Comment by PettitWesley** [comment]
@agup006 @patrick-stephens Do we want to get this merged?

**Comment by patrick-stephens** [comment]
Seems reasonable to me as best practice: `exec` is not possible as an input plugin for (production) containers but there are other uses as indicated and my understanding is it should not affect existing behaviour as I understand it. I'll trigger integration tests though.

**Comment by rittneje** [comment]
@edsiper @niedbalski @koleini @fujimotos @nokute78

**Comment by niedbalski** [comment]
@PettitWesley can you approve this PR, then?

**Comment by rittneje** [comment]
@edsiper @niedbalski @koleini @fujimotos @nokute78 @PettitWesley

**Comment by rittneje** [comment]
@edsiper @niedbalski @koleini @fujimotos @nokute78 @PettitWesley

**Review by PettitWesley — APPROVED** [review]

**Comment by rittneje** [comment]
@edsiper @niedbalski @koleini @fujimotos @nokute78

**Comment by rittneje** [comment]
@edsiper @niedbalski @koleini @fujimotos @nokute78

**Comment by rittneje** [comment]
@edsiper @niedbalski @koleini @fujimotos @nokute78

**Comment by rittneje** [comment]
@edsiper @niedbalski @koleini @fujimotos @nokute78

**Comment by rittneje** [comment]
@edsiper @niedbalski @koleini @fujimotos @nokute78

**Comment by rittneje** [comment]
@edsiper @niedbalski @koleini @fujimotos @nokute78

**Comment by rittneje** [comment]
@edsiper @niedbalski @koleini @fujimotos @nokute78

**Comment by rittneje** [comment]
@edsiper @niedbalski @koleini @fujimotos @nokute78

**Comment by rittneje** [comment]
@edsiper @niedbalski @koleini @fujimotos @nokute78 I don't know what to make of the integration test failures, since they don't provide any useful output. Also, this PR has been open for almost 10 months with no feedback. Can you review it?

**Comment by rittneje** [comment]
@edsiper It's pretty disheartening that this was ignored for 10 months. I will definitely think twice before attempt to contribute to this repo again.

**Pull request closed without merging**
